### PR TITLE
HPCC-21077 Add a statistic for graph execution count

### PR DESCRIPTION
**Pull request by jakesmith**
Track and publish number of times each graph executes.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: system/jlib/jstatcodes.h [hunk 1]
```diff
@@ -213,6 +213,7 @@ enum StatisticKind
     StNumSmartJoinDegradedToLocal,      // number of times global smart join degraded to local smart join (<=1 unless in loop)
     StNumSmartJoinSlavesDegradedToStd,  // number of times a slave in smart join degraded from local smart join to standard hash join
     StNumAttribsSimplifiedTooComplex,
+    StNumGraphExecutions,               // number of times graph executed (<=1 unless loop or child query graph)
```
Comment (ghalliday) [inline system/jlib/jstatcodes.h]: Ideally we should have a limited number of different St kinds, and the scope would indicate that it applied to a graph.  So this should be reusing StNumExecutions (or NumIterations /NUmStarted/NumStopped). The scope indicates that it applies to a graph, so shouldn't be in the stat name.

File: thorlcr/graph/thgraphmaster.ipp [hunk 2]
```diff
@@ -35,6 +35,101 @@
...
+{
+protected:
+    CJobBase &ctx;
+    unsigned __int64 max, min, tot, avg;
+    unsigned maxSkew, minSkew, minNode, maxNode;
+    UInt64Array counts;
+    StatisticKind kind;
+
+public:
+    IMPLEMENT_IINTERFACE;
```
Comment (ghalliday) [inline thorlcr/graph/thgraphmaster.ipp]: not new, but not needed.

**Review by ghalliday — COMMENTED** [review]
@jakesmith a couple of minor comments

File: system/jlib/jstatcodes.h (hunk 1 above)
Comment (jakesmith) [reply system/jlib/jstatcodes.h]: ok, I'll change it to StNumExecutions

File: thorlcr/graph/thgraphmaster.ipp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphmaster.ipp]: agreed, will remove.

**Comment by jakesmith** [comment]
Have made changed and squashed.
NB: NumExecutions already existed as a defined property, so reused it.

@ghalliday - Please review

**Comment by AttilaVamos** [comment]
As I told several times: We have build time on-fly download/build stuff without any proper error handling/failure tolerance in our (c)make scripts. Sometimes the download generates a zero length file and the further build failed on it.

**Pull request merged**